### PR TITLE
Debugger: support address prefixes for slot, bank, language card & ROM

### PR DESCRIPTION
**Pull request by tomcw**
PR for #451

### CODE REVIEW

File: source/Debugger/Debugger_Assembler.cpp [hunk 1]
```diff
@@ -1215,7 +1215,7 @@ bool AssemblerGetArgs( int iArg, int nArgs, WORD nBaseAddress )
 		{
 		}
 		else
-		if (iToken == TOKEN_SEMI) // comment
```
Comment (Michaelangel007) [inline source/Debugger/Debugger_Assembler.cpp:1218]: Thanks for this catch! N.B. The original intent was to support EOL comments in assembly language via the `;` character.  Supporting the debugger's use of C++'s `//` is fine but we probably should also support the original semi-colon as `TOKEN_COMMENT_EOL_2`.

Maybe rename `TOKEN_COMMENT_EOL` to `TOKEN_COMMENT_EOL_1` ?

Comment (Michaelangel007) [reply source/Debugger/Debugger_Assembler.cpp:1218]: Ah, using `//` for divide floor is probably a better usage.

File: source/Debugger/Debugger_Range.cpp [hunk 2]
```diff
@@ -34,31 +34,36 @@ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 
 
 //===========================================================================
-bool Range_GetPrefix(const int iArg, BYTE& slot, BYTE& bank, BYTE& lc, bool& isROM)
+RangeType_t Range_GetPrefix(const int iArg, BYTE& slot, BYTE& bank, BYTE& lc, bool& isROM)
```
Comment (Michaelangel007) [inline source/Debugger/Debugger_Range.cpp]: 👍 for updating the return type

File: source/Debugger/Debugger_Range.cpp [hunk 3]
```diff
@@ -73,7 +73,7 @@ RangeType_t Range_GetPrefix(const int iArg, Breakpoint_t* pBP)
 	{
 		if (g_aArgs[iArg].nValue > 0xff)
 		{
-			ConsoleDisplayError("Address prefix bad: Use bank 0 - FF (or 0 - 7 for Saturn).");
+			ConsoleDisplayError("Address prefix bad: Use bank 0-FF (or 0-7 for Saturn).");
```
Comment (Michaelangel007) [inline source/Debugger/Debugger_Range.cpp]: This error should be more descriptive. Also, we probably want to markup the hex numbers with `CHC_NUMBER`.
i.e.
```c
"Address prefix out-of-range. Valid slots are " CHC_NUMBER "0" CHC_OPERATOR "-" CHC_NUMBER "7" CHC_DEFAULT "."
```

File: source/Debugger/Debug.cpp [hunk 4]
```diff
@@ -136,6 +136,7 @@ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 	};
 
 	static WORD g_uBreakMemoryAddress = 0;
+	static int g_breakpointHitID = -1;
```
Comment (Michaelangel007) [inline source/Debugger/Debug.cpp]: Does this need to be reset in `DebugInitialize()` ?

File: source/Debugger/Debug.cpp [hunk 5]
```diff
@@ -2087,6 +2088,69 @@ void _BWZ_EnableDisableViaArgs ( int nArgs, Breakpoint_t * aBreakWatchZero, cons
 	}
 }
 
+//===========================================================================
+static std::string GetFullPrefixAddrForBreakpoint(const Breakpoint_t* pBP, WORD address, bool padding)
+{
+	char szSlot[] = "sN/";
+	char szBank[] = "NN/";
+	char szLangCard[] = "lN/";
```
Comment (Michaelangel007) [inline source/Debugger/Debug.cpp:2137]: Using `s` prefix for strings if sufficient.  We don't need to use `z` as that is redundant and too verbose.
```c
	char sSlot    [] = "sN/"; // Saturn Slot
	char sBank    [] = "NN/"; // Ramworks Bank
	char sLangCard[] = "lN/"; // Language Card
```

Comment (Michaelangel007) [reply source/Debugger/Debug.cpp:2137]: Function should have comment linking to GitHub Issue for syntax.

```c
// For BP syntax
// See:
// * https://github.com/AppleWin/AppleWin/issues/451
```

**Comment by Michaelangel007** [comment]
@tomcw Some minor notes above but looking good!

Review progress:

* [x] [1f38cd8](https://github.com/AppleWin/AppleWin/pull/1419/commits/1f38cd800d8a3f90323402ad4a54531e71f1a9b3)
* [x] [e2f56ca](https://github.com/AppleWin/AppleWin/pull/1419/commits/e2f56caffe52298fb44780250003a7a1ca0868de)
* [x] [d0b39ae](https://github.com/AppleWin/AppleWin/pull/1419/commits/d0b39ae9aa00fd8aec774d1f806a5e9c412b4a51)
* [x] [0355271](https://github.com/AppleWin/AppleWin/pull/1419/commits/0355271adf397a0c549a78b171465cc51a2c6c40)
* [x] [68639d7](https://github.com/AppleWin/AppleWin/pull/1419/commits/68639d78f4c710215c812277d5262e2d6f8dcc59)
* [x] [6f0670e](https://github.com/AppleWin/AppleWin/pull/1419/commits/6f0670ed228eb33f00107aedf77512fda1532d2e)
* [x] [9fb2383](https://github.com/AppleWin/AppleWin/pull/1419/commits/9fb238368da7a29e744055422aad3ef36f511fff)
* [x] [fc8bc20](https://github.com/AppleWin/AppleWin/pull/1419/commits/fc8bc20737a47031ae36ca642da8867625e7193d)
* [x] [f508a11](https://github.com/AppleWin/AppleWin/pull/1419/commits/f508a118e7762ab1fc1c08e71eea7f630e9d3123)
* [x] [1fb5be3](https://github.com/AppleWin/AppleWin/pull/1419/commits/1fb5be339c12d8de16a8e6bd3054ef8e8218080f)
* [x] [00c2f9b](https://github.com/AppleWin/AppleWin/pull/1419/commits/00c2f9b8643a9a1290999674f9c1e071aab231df)
* [x] [1584809](https://github.com/AppleWin/AppleWin/pull/1419/commits/1584809d9e34475d6fb2c2d03e67b09d20cb3edc)
* [x] [5b75d51](https://github.com/AppleWin/AppleWin/pull/1419/commits/5b75d5149d61300753245e95989d025a5cdfb035)
* [x] [b114c1d](https://github.com/AppleWin/AppleWin/pull/1419/commits/b114c1df4d9c9460b0d7e4d2aabc8cf1773e7c67)
* [x] [1684ab3](https://github.com/AppleWin/AppleWin/pull/1419/commits/1684ab3c1cc24688f287daefced349e790fb2842)
* [x] [2e65b39](https://github.com/AppleWin/AppleWin/pull/1419/commits/2e65b3935077aa0f0bed2a359214136dc4e61d60)
* [x] [97cb2db](https://github.com/AppleWin/AppleWin/pull/1419/commits/97cb2db5b343e9bfb7691c93ccc6911ddee910a4)
* [x] [804665f](https://github.com/AppleWin/AppleWin/pull/1419/commits/804665f865cc3b363905a8864ab5c97c01ecb401)
* [x] [a70529f](https://github.com/AppleWin/AppleWin/pull/1419/commits/a70529f7804338117f53ff60095cdb16263ec8b9)

File: source/Debugger/Debugger_Assembler.cpp (hunk 1 above)
Comment (tomcw) [reply source/Debugger/Debugger_Assembler.cpp:1218]: NB, I support the special case of lines starting with `//` as a comment too.
See `Debugger_Parser.cpp`: `ArgsGet()`.

For example, this is used in your `A2_DOS33.SYM2` [here](https://github.com/AppleWin/AppleWin/blob/master/bin/A2_DOS33.SYM2).

File: source/Debugger/Debugger_Range.cpp (hunk 3 above)
Comment (tomcw) [reply source/Debugger/Debugger_Range.cpp]: I've changed all 3 from "Address prefix bad: ..." to "Address prefix out-of-range. ...".

btw, `CHC_NUMBER` only exists in a `#if 0` block of code (in Debugger_Console.h) and `CHC_OPERATOR` doesn't exist.

File: source/Debugger/Debug.cpp (hunk 4 above)
Comment (tomcw) [reply source/Debugger/Debug.cpp]: Yes - good catch.

File: source/Debugger/Debug.cpp [hunk 6]
```diff
@@ -1183,102 +1183,97 @@ bool _CheckBreakpointValueWithPrefix(Breakpoint_t* pBP, int nVal)
...
-			else if (0x2000 < nVal && nVal <= 0x3FFF && ((GetMemMode() & (MF_80STORE | MF_HIRES)) == (MF_80STORE | MF_HIRES)))
-			{
-				if ((pBP->bank == 0x00 && !(GetMemMode() & MF_PAGE2))
-					|| (pBP->bank != 0x00 && (GetMemMode() & MF_PAGE2)))
-				{
-					bStatus = true;
-				}
-			}
-			else if (nVal < 0xC000)
+			if (noRamworksOrSaturnBank
```
[the remark this answers is not in the document]
Comment (tomcw) [reply source/Debugger/Debug.cpp]: Updated.

**Comment by Michaelangel007** [comment]
Looking good. Thanks!

**Comment by Michaelangel007** [comment]
Reviewed:

* [x] https://github.com/AppleWin/AppleWin/pull/1419/commits/8ee669db8e9e2d5631ade7941fec078efadf7a52
* [x] https://github.com/AppleWin/AppleWin/pull/1419/commits/3a5993e996f91b8813c2922ead14b6f6d3d53383

Do we need to add a blurb in the help for WHY the `/` division operator has changed? i.e. Context for people.

e.g.

/!\ Older versions of AppleWin's Debugger used `/` for (integer) division. The debugger needed a way for users to refer to auxiliary memory and banked memory addresses. The IIgs monitor used `/` for banks and `_` for division. i.e. `0/0000` (main) and `1/0000` aux.  Since division is rarely used we changed this to be in-line with the IIgs native syntax. This way you can specify a breakpoint on a Language Card memory address, etc.

**Comment by tomcw** [comment]
>Do we need to add a blurb in the help for WHY the `/` division operator has changed? i.e. Context for people.

Yes, I agree. Once we settle on `//` or `_` then I'll extend the help to include some extra info about this.

**Comment by tomcw** [comment]
@Michaelangel007, @sicklittlemonkey: final call for comments on this PR.
Primarily it's about using address prefixes to allow breakpoints to be set on main/aux and different LC 4K banks.
But as a (minor contraversial) side effect, the rarely-used divide operator changed from `/` to `//`.

btw, I've added some extensive regression tests here:
https://github.com/AppleWin/AppleWin-Test/blob/master/Projects/GH451-Debugger-prefix-breakpoints/Debugger-prefix-breakpoints.a
(Sadly all manual at the moment, but I have ideas to automate them)

**Comment by Michaelangel007** [comment]
**TL:DR;** LGTM! Also, added indicators for LC1, LC2, Aux, and Aux N banks.

One of these days I'll remember all these git commands ...

```
git fetch --all
git checkout origin/GH451-Debugger_addr-prefix
git checkout GH451-Debugger_addr-prefix
git branch --set-upstream-to=origin/GH451-Debugger_addr-prefix GH451-Debugger_addr-prefix
```

I did a quick test:

```asm
        ORG $300

        LDA $C08B          ; Bank 1
        LDA $C08B          ; Writeable

        LDY #LC_0 - LC_1-1 ; Copy LC_1..LC_0
Copy    LDA  LC_1,Y
        STA $D000,Y        ; to Bank 1
        DEY
        BPL Copy

        LDA $C088          ; Run from Bank 1
        JSR $D000
        LDA $C082
        LDA $C082
        RTS
LC_1    LDA #$C1  ; Bank 1
        STA $0400
        RTS
LC_0
```

To run:
```
BLOAD "TEST",300
BPX 300
BPX 317
BPX LC1/D002
<F7>
CALL 768
```

And we hit the bp in LC1.  VERY nice.

~~**For future work** we have a space after the `=`~~ We have space which I believe we can hijack to show the LC1/2, Aux/Main with a single glyph.  The bookmark glyphs for (1) and (2) will work nicely here. 

```
B0 PC=①D001   (LC Bank 1)
B0 PC=②D002   (LC Bank 2)
B0 PC=AD00A   (Aux. Mem)
```

Ah, we can't use `=` as that is reserved for 2 characters in `g_aBreakpointSymbols[]`;

Moved the indicator to be after the breakpoint numbers since we show this in hex:

```
B0①PC= D001   (LC Bank 1)
B1②PC= D002   (LC Bank 2)
B2αPC= A001   (Aux. Mem.)
B3♦PC= A002   (RW Bank > 1)
```
~~We'll have to use an lower-a~~ and a different color for Aux. I've used `α` as a temp. glyph for emphasis. See the screenshot in #451.

**Comment by Michaelangel007** [comment]
@tomcw  ~~I'll add this in a future PR but if you wanted to play with the updated display...~~ I've added the bank indictors.  See below. Q. Looks like `nBank` will be set to 1 for aux. memory?

```
diff --git a/source/Debugger/Debugger_Display.cpp b/source/Debugger/Debugger_Display.cpp
index 5bb0111e..bd1744cc 100644
--- a/source/Debugger/Debugger_Display.cpp
+++ b/source/Debugger/Debugger_Display.cpp
@@ -1077,7 +1077,15 @@ void DrawBreakpoints ( int line )
 
 			DebuggerSetColorBG( DebuggerGetColor( BG_INFO ));
 			DebuggerSetColorFG( DebuggerGetColor( FG_INFO_BULLET ) );
-			PrintTextCursorX( StrFormat("%X ", iBreakpoint).c_str(), rect2);
+			PrintTextCursorX( StrFormat("%X", iBreakpoint).c_str(), rect2);
+
+			// If the BP is in LC1 or LC2 display a bookmark symbol (1) or (2)
+			if (pBP->nLangCard != -1)
+			{
+				FillRect( GetDebuggerMemDC(), &rect2, g_hConsoleBrushBG );
+				PrintGlyph( rect2.left, rect2.top, 0x80 + pBP->nLangCard ); // Glyphs 0x80 .. 0x89 = Unicode U+24EA, U+2460 .. U+2468
+			}
+			rect2.left += g_aFontConfig[ FONT_DISASM_DEFAULT ]._nFontWidthAvg;
 
 //			DebuggerSetColorFG( DebuggerGetColor( FG_INFO_OPERATOR ) );
 //			PrintTextCursorX( ".", rect2 );
```

**Comment by Michaelangel007** [comment]
What the LC1/2 breakpoint indicator now looks like:

image

**Comment by Michaelangel007** [comment]
Added Aux.1 and Aux.N indicators

image

**Comment by tomcw** [comment]
@Michaelangel007 - what if it's LC1/2 _and_ aux?

`bpm 1/lc1/d001`
`bpm 2/lc2/d001`

I was thinking we could use a mouse-over tooltip to show the full breakpoint.
(Or just make the screen bigger!)

---

btw, we have an extra char at the end:
```
B0 M = D001:DFFFR 
                 ^
```

So you could move your LC1/2 to the end of the line:

bpm l1/d000
bpm l2/d000,1000
bpmr l1/d000
bpmw l2/d000,1000

And extend the RamWorks bank to show (1)..(9) then "diamond":
bpm 1/10
bpm 2/20
bpm 9/90
bpm 10/100

Both:
bpm 1/l1/d000
bpm 2/l2/d000,1000
bpmr 9/l1/d000
bpmw 10/l2/d000,1000

BPs

**Comment by Michaelangel007** [comment]
OOF! I _knew_ I was forgetting an edge case or two.

I like using the extra room on the end. We'll probably need to right justify it so it looks good.

You got some great test cases -- I'll play around with those and see what look good.

I'm not sure if we can make the screen bigger though?

A mouse tool-tip is probably still required. I need to see if we have that functionality.

I'm happy to merge this PR and then spin-off a new PR 

OR 

we can create a sub-task: Debugger UI Bank BP cleanup.

**Comment by tomcw** [comment]
>We'll probably need to right justify it so it looks good.

Agreed... do this, then merge it.

**Comment by Michaelangel007** [comment]
@tomcw I've had a chance to noodle on this.  I don't think splitting up the slot/bank is a good idea. We have room to show both together plus it is more logical to keep them together:

Setting `DEBUG_FORCE_DISPLAY` here is an example:

**Before:**

before_all_symbols

**After:**

after_all_symbols

I nudged the:

* BP left column over 1/2 a glyph to the left, `INFO_COL_2`
* added 1/2 glyph spacer before the Slot
* added 1/2 glyph spacer after the Bank
* Added `DEBUG_FORCE_DISPLAY` support to display dummy values

**Comment by Michaelangel007** [comment]
Example script:

```
BPC *
BPM 0
BPM LC1/D010
BPM LC2/D020
BPM 1/0
BPM 1/LC1/D110
BPM 1/LC2/D120
BPM 2/2
BPM 3/LC1/D210
BPM 4/LC2/D220
BPM 5/5
BPM 6/LC1/D610
BPM 7/LC2/D720
BPM 8/8
BPM 9/LC1/D910
BPM 0A/LC2/DA20
BPM B/0B
```

bp_slot_lc_example

**Comment by tomcw** [comment]
@Michaelangel007 - you'll need to shift it left by 1/2 a glyph, as the "R" or "W" final char is half off the screen, eg:
bpmr l2/d000,1000
bpmw l2/d000,1000

(Otherwise it looks good)

**Comment by Michaelangel007** [comment]
@tomcw Thanks for the catch!  I changed the spacer before and after to be 2 pixels which is BARELY enough room for this to fit.

I also fixed up the bg/fg colors since the `R` and `W` were unreadable (they blended into the end range address) and added the 4 missing debuggers colors to the color scheme.

image

**Comment by Michaelangel007** [comment]
I added a proper 3x5 mini hex font so we can display the hex bank.

image

```
bpc *
bpm 00/l1/d000,1000
bpmr 12/l2/d000,1000
bpmw 34/l2/d000,1000
bpm 1/A000
```

image

**Comment by Michaelangel007** [comment]
**Note:** I _didn't_ track down the correct GIMP settings to export the `Debug_Font.bmp` properly -- just pasted the touch up into mspaint and saved it.

**Comment by tomcw** [comment]
I've added support for RamWorks bank 0x100 too.
(RamWorks _fully-populated_ can be 0x100 * 64K = 16MB, so Apple //e in total will have 16M+64K bytes of memory)

```
bpmw 100/l1/d000,800
```

**Comment by tomcw** [comment]
BP-bank-100

```
bpc *
bpm 00/l1/d000,1000
bpmr 12/l2/d000,1000
bpmw 34/l2/d000,1000
bpm 1/A000
bpmw 100/l1/d000,800
```

**Comment by Michaelangel007** [comment]
Ah, I was wondering why there was a bigger gap. :-)

I'm happy with this.  The one minor question is do we want to touch up the "extra" red background spacer on the right? i.e. End space on the `A000` line.

**Comment by tomcw** [comment]
Yes, I think it'll look neater to tidy (trim) it.

**Comment by Michaelangel007** [comment]
Argh, I think I messed up the git history.

Last commit on this branch before my fix was b9da2dab3e36a3317

**Comment by tomcw** [comment]
I've still got the local changes, and the commits still exist... I wonder what's the best way to restore this?

**Comment by Michaelangel007** [comment]
> I've still got the local changes, and the commits still exist... I wonder what's the best way to restore this?

Yeah, the commits aren't lost.

I _think_ maybe you force push your last commit to the branch?

`git push --force`

I'm not sure why `Reopen and comment` is grayed out?

**Comment by tomcw** [comment]
>I'm not sure why `Reopen and comment` is grayed out?

"There are no new commits in the GH451-Debugger_addr-prefix branch"

**Comment by Michaelangel007** [comment]
You want to hop on a zoom call?

**Comment by tomcw** [comment]
$ git push --force
Enumerating objects: 265, done.
Counting objects: 100% (265/265), done.
Delta compression using up to 16 threads
Compressing objects: 100% (246/246), done.
Writing objects: 100% (246/246), 37.95 KiB | 1.81 MiB/s, done.
Total 246 (delta 197), reused 0 (delta 0), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (197/197), completed with 19 local objects.
To https://github.com/AppleWin/AppleWin.git
 + 3f3bf8ed...b9da2dab GH451-Debugger_addr-prefix -> GH451-Debugger_addr-prefix (forced update)

**Comment by Michaelangel007** [comment]
Hmm, is this middle commit correct?

https://github.com/AppleWin/AppleWin/commits/GH451-Debugger_addr-prefix/

* [x] b9da2dab3e36a33172280c1e0f8d4bbdd68f12e2
* [ ] b0850fa63679c2ae8bc0d79438df38a57e31adf2
* [x] f3360e726a0b19bfadda4843744a5696cf0f3846

**Comment by tomcw** [comment]
https://github.com/AppleWin/AppleWin/commit/b0850fa63679c2ae8bc0d79438df38a57e31adf2 is a legitimate commit from me that's in this branch.

**Comment by Michaelangel007** [comment]
Ok, I think we are good now.

Whew!

**Comment by tomcw** [comment]
Yes - well done! Looks like you restored it 👍

**Comment by Michaelangel007** [comment]
I'm _still_ not exactly sure why the Reopen button became active. Maybe my last commit re-enabled it?

You want to do a quick sanity check/test?

**Comment by Michaelangel007** [comment]
image

**Comment by tomcw** [comment]
>You want to do a quick sanity check/test?

All fine 👍
